libkbfs/fetch_decider.go

```diff
@@ -0,0 +1,135 @@
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"sync"
+	"time"
+
+	"github.com/keybase/client/go/logger"
+	"golang.org/x/net/context"
+)
+
+// fetchDecider is a struct that helps avoid having too frequent calls
+// into a remote server.
+type fetchDecider struct {
+	log     logger.Logger
+	fetcher func(ctx context.Context) error
+	tagKey  interface{}
+	tagName string
+	clock   clockGetter
+
+	blockingForTest chan<- struct{}
+
+	lock    sync.Mutex
+	readyCh chan struct{}
+	errPtr  *error
+}
+
+func newFetchDecider(
+	log logger.Logger, fetcher func(ctx context.Context) error,
+	tagKey interface{}, tagName string,
+	clock clockGetter) *fetchDecider {
+	return &fetchDecider{
+		log:     log,
+		fetcher: fetcher,
+		tagKey:  tagKey,
+		tagName: tagName,
+		clock:   clock,
+	}
+}
+
+func (fd *fetchDecider) launchBackgroundFetch(ctx context.Context) (
+	readyCh <-chan struct{}, errPtr *error) {
+	fd.lock.Lock()
+	defer fd.lock.Unlock()
+
+	if fd.readyCh != nil {
+		fd.log.CDebugf(ctx, "Waiting on existing fetch")
+		// There's already a fetch in progress.
+		return fd.readyCh, fd.errPtr
+	}
+
+	fd.readyCh = make(chan struct{})
+	fd.errPtr = new(error)
+
+	id, err := MakeRandomRequestID()
+	if err != nil {
+		fd.log.Warning("Couldn't generate a random request ID: %v", err)
+	}
+	fd.log.CDebugf(
+		ctx, "Spawning fetch in background with tag:%s=%v", fd.tagName, id)
+	go func() {
+		// Make a new context so that it doesn't get canceled
+		// when returned.
+		logTags := make(logger.CtxLogTags)
+		logTags[fd.tagKey] = fd.tagName
+		bgCtx := logger.NewContextWithLogTags(
+			context.Background(), logTags)
+		bgCtx = context.WithValue(bgCtx, fd.tagKey, id)
+		// Make sure a timeout is on the context, in case the
+		// RPC blocks forever somehow, where we'd end up with
+		// never resetting backgroundInProcess flag again.
+		bgCtx, cancel := context.WithTimeout(bgCtx, 10*time.Second)
+		defer cancel()
+		err := fd.fetcher(bgCtx)
+
+		// Notify everyone we're done fetching.
+		fd.lock.Lock()
+		defer fd.lock.Unlock()
+		fd.log.CDebugf(bgCtx, "Finished fetch: %+v", err)
+		*fd.errPtr = err
+		readyCh := fd.readyCh
+		fd.readyCh = nil
+		fd.errPtr = nil
+		close(readyCh)
+	}()
+	return fd.readyCh, fd.errPtr
+}
+
+// Do decides whether to block on a fetch, launch a background fetch
+// and use existing cached value, or simply use the existing cached
+// value with no more fetching. The caller can provide a positive
+// tolerance, to accept stale LimitBytes and UsageBytes data. If
+// tolerance is 0 or negative, this always makes a blocking call using
+// `fd.fetcher`.
+//
+// 1) If the age of cached data is more than blockTolerance, it blocks
+// until a new value is fetched and ready in the caller's cache.
+// 2) Otherwise, if the age of cached data is more than bgTolerance,
+// a background RPC is spawned to refresh cached data using `fd.fetcher`,
+// but returns immediately to let the caller use stale data.
```

Comment:
Yeah, the point is to try to refresh the value before `blockTolerance` is hit, so that future callers don't ever block while waiting.

(This is the behavior that `EventuallyConsistentQuotaUsage` has always had, I just factored it out here to be reusable.)